pkg/middleware/instrumentation.go

```diff
@@ -0,0 +1,120 @@
+
+		// The latency of all requests bucketed by HTTP method
+		requestLatency := RegisterRequestsLatencyHistogram(registerer)
+
+		return http.HandlerFunc(func(wr http.ResponseWriter, req *http.Request) {
+			// Skip recording metrics for the metrics endpoint path
+			// The handler for the metrics endpoint will record those requests
+			// separately.
+			if req.URL.EscapedPath() == metricsPath {
+				next.ServeHTTP(wr, req)
+			} else {
+				promhttp.InstrumentHandlerCounter(requestsCounter,
+					promhttp.InstrumentHandlerInFlight(inFlightRequests,
+						promhttp.InstrumentHandlerDuration(requestLatency, next),
+					),
+				)(wr, req)
+			}
+		})
+	}
+}
+
+func RegisterRequestsCounter(registerer prometheus.Registerer) *prometheus.CounterVec {
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "oauth2_proxy_requests_total",
+			Help: "Total number of requests by HTTP status code.",
+		},
+		[]string{"code"},
+	)
+
+	// Initialize likely HTTP status codes.
+	counter.WithLabelValues("200") // OK
+	counter.WithLabelValues("202") // Accepted
+	counter.WithLabelValues("204") // NoContent
+	counter.WithLabelValues("302") // Found
+	counter.WithLabelValues("400") // BadRequest
+	counter.WithLabelValues("401") // Unauthorized
+	counter.WithLabelValues("403") // Forbidden
+	counter.WithLabelValues("404") // NotFound
+	counter.WithLabelValues("408") // RequestTimeout
+	counter.WithLabelValues("500") // InternalServerError
+	counter.WithLabelValues("502") // BadGateway
+	counter.WithLabelValues("504") // GatewayTimeout
+
+	if err := registerer.Register(counter); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			counter = are.ExistingCollector.(*prometheus.CounterVec)
+		} else {
+			panic(err)
+		}
+	}
+
+	return counter
+}
+
+func RegisterInflightRequestsGauge(registerer prometheus.Registerer) prometheus.Gauge {
+	gauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "oauth2_proxy_requests_in_flight",
+		Help: "Current number of requests being served.",
+	})
+
+	if err := registerer.Register(gauge); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			gauge = are.ExistingCollector.(prometheus.Gauge)
+		} else {
+			panic(err)
+		}
+	}
+
+	return gauge
+}
+
+func RegisterRequestsLatencyHistogram(registerer prometheus.Registerer) *prometheus.HistogramVec {
+	histogram := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "oauth2_proxy_response_duration_seconds",
+			Help:    "A histogram of request latencies.",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"method"},
+	)
+
+	if err := registerer.Register(histogram); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			histogram = are.ExistingCollector.(*prometheus.HistogramVec)
+		} else {
+			panic(err)
+		}
+	}
+
+	return histogram
+}
+
+func MetricsServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	promhttp.Handler().ServeHTTP(rw, req)
+}
```

Comment:
It will display the formatted metrics but it would only display metrics register with the default registry. So yeah, missing our custom metrics but currently it is only using default registry. Without storing the registry somewhere and passing it here we wouldn't be able to display them.

Since we are moving this to a different port any how, I'll push some things into `oauthproxy.go` and wire a handler for the custom registry.

Reply:
I might suggest doing as much as you can in a `metrics.go` file rather than adding to the huge `oauthproxy.go` file